src/Razor/test/Microsoft.AspNetCore.Razor.LanguageServer.Test/CodeActions/CodeActionEndToEndTest.cs

```diff
@@ -0,0 +1,720 @@
+            createRazorCodeActionResolversFn: () => new[] { new GenerateMethodCodeActionResolver(DocumentContextFactory) },
+            diagnostics: diagnostics);
+    }
+
+    [Theory]
+    [InlineData("[||]Exists")]
+    [InlineData("E[||]xists")]
+    [InlineData("Exists[||]")]
+    public async Task Handle_GenerateMethod_Method_ExistsInCodeBlock(string cursorAndMethodName)
+    {
+        var input = $$"""
+            <button @onclick="{{cursorAndMethodName}}"></button>
+            @code {
+                public void Exists()
+                {
+                }
+            }
+            """;
+
+        TestFileMarkupParser.GetSpan(input, out input, out var textSpan);
+        var razorFilePath = "file://C:/path/test.razor";
+        var codeDocument = CreateCodeDocument(input, filePath: razorFilePath);
+        var razorSourceText = codeDocument.GetSourceText();
+        var uri = new Uri(razorFilePath);
+        await CreateLanguageServerAsync(codeDocument, razorFilePath);
+        var documentContext = CreateDocumentContext(uri, codeDocument);
+        var requestContext = new RazorRequestContext(documentContext, Logger, null!);
+
+        var result = await GetCodeActionsAsync(uri, textSpan, razorSourceText, requestContext, razorCodeActionProviders: new[] { new GenerateMethodCodeActionProvider() });
+        Assert.False(result.Where(e => ((RazorVSInternalCodeAction)e.Value!).Title == "Generate 'DoesNotExist' Method").Any());
+    }
+
+    [Theory]
+    [InlineData("")]
+    [InlineData("\r\n")]
+    [InlineData("\r\npublic void Exists(){}\r\n")]
+    public async Task Handle_GenerateMethod_CodeBehindFile_Exists(string spacingOrMethod)
+    {
+        var input = """
+            <button @onclick="[||]DoesNotExist"></button>
+            """;
+
+        var expectedRazorContent = """
+            <button @onclick="DoesNotExist"></button>
+            """;
+
+        var initialCodeBehindContent = $$"""
+            namespace __GeneratedComponent
+            {
+                public partial class test
+                {{{spacingOrMethod}}
+                }
+            }
+            """;
+
+        var expectedCodeBehindContent = $$"""
+            namespace __GeneratedComponent
+            {
+                public partial class test
+                {{{spacingOrMethod}}
+                    private void DoesNotExist()
+                    {
+                        throw new NotImplementedException();
+                    }
+                }
+            }
+            """;
+
+        var razorFilePath = FilePathNormalizer.Normalize($"{Path.GetTempPath()}test.razor");
+        var codeBehindFilePath = FilePathNormalizer.Normalize($"{Path.GetTempPath()}test.razor.cs");
+        var diagnostics = new[] { new Diagnostic() { Code = "CS0103", Message = "The name 'DoesNotExist' does not exist in the current context" } };
+
+        TestFileMarkupParser.GetSpan(input, out input, out var textSpan);
```

Comment:
Love these tests! Is it possible to extract the common code in this test, and the next one, so there is a new `VerifyXXX` method that handles code behind files too? Could definitely see this being useful in future.